### ISSUE TITLE
Calling `Tensor::backward()` inside a custom op's `Backward::backward` deadlocks permanently

### ISSUE THREAD

**Issue by laurigates** [tracel-ai/burn#5193]
## Summary

A custom autodiff op whose `Backward::backward` implementation runs an inner `Tensor::backward()` — the natural shape for op-level gradient checkpointing (recompute a region and backprop it during the outer backward) — hangs forever, even when the inner graph is a brand-new, fully disjoint graph sharing no tensors with the outer one.

Two pieces of `burn-autodiff`'s locking interact:

1. **The outer backward executes every step while holding its graph's state mutex.** [`GraphMutexClient::backward` wraps the whole `server.backward` step execution in `graph.state.lock()`](https://github.com/tracel-ai/burn/blob/e5467f02c3cf88eb5d709f190c170005ce26038d/crates/burn-autodiff/src/runtime/graph.rs#L116-L129), so user `Backward::backward` code runs with that lock held.
2. **Every `backward()` call ends with a global sweep that locks every registered graph.** [`GraphCleaner::cleanup_orphaned_entries` iterates `STATE.graphs` and takes each graph's `state.lock()`](https://github.com/tracel-ai/burn/blob/e5467f02c3cf88eb5d709f190c170005ce26038d/crates/burn-autodiff/src/runtime/graph.rs#L138-L159) — including the outer graph whose mutex the current thread already holds.

`state` is a non-reentrant [`parking_lot::Mutex`](https://github.com/tracel-ai/burn/blob/e5467f02c3cf88eb5d709f190c170005ce26038d/crates/burn-autodiff/src/runtime/graph.rs#L65), so the inner backward's cleanup sweep self-deadlocks on the outer graph's lock. Moving the inner backward to a worker thread only converts it into a cross-thread deadlock (the outer thread blocks on the join while the worker blocks on the outer mutex).

Reproduced by executing the program below against v0.21.0 (same structure at [`graph.rs#L114-L146`](https://github.com/tracel-ai/burn/blob/v0.21.0/crates/burn-autodiff/src/runtime/graph.rs#L114-L146) and [`#L153-L177`](https://github.com/tracel-ai/burn/blob/v0.21.0/crates/burn-autodiff/src/runtime/graph.rs#L153-L177)); the same hold-lock-across-step + lock-every-graph-in-cleanup pattern is present at current `main` (`e5467f0`, links above — code-inspected, not executed).

## Reproduction (burn 0.21.0, ndarray)

```toml
[package]
name = "nested-backward-repro"
version = "0.1.0"
edition = "2021"

[dependencies]
burn = { version = "0.21.0", features = ["autodiff", "ndarray"] }
```

```rust
use burn::backend::autodiff::checkpoint::base::Checkpointer;
use burn::backend::autodiff::grads::Gradients;
use burn::backend::autodiff::ops::{Backward, Ops, OpsKind, unary};
use burn::backend::{Autodiff, NdArray};
use burn::tensor::{Tensor, TensorPrimitive};

type Inner = NdArray;
type Ad = Autodiff<Inner>;

/// Forward = identity; backward runs a fully disjoint inner autodiff graph
/// (the natural shape of op-level gradient checkpointing).
fn nested_identity(x: Tensor<Ad, 1>) -> Tensor<Ad, 1> {
    #[derive(Debug)]
    struct NestedIdentity;

    impl Backward<Inner, 1> for NestedIdentity {
        type State = ();

        fn backward(
            self,
            ops: Ops<Self::State, 1>,
            grads: &mut Gradients,
            _checkpointer: &mut Checkpointer,
        ) {
            unary::<Inner, _>(ops.parents, ops.node, grads, |grad| {
                eprintln!("entering nested inner backward…");
                // A brand-new graph: fresh tensors, no connection to the
                // outer graph whatsoever.
                let a = Tensor::<Ad, 1>::ones([4], &Default::default()).require_grad();
                let inner_loss = a.clone().sum();
                let inner_grads = inner_loss.backward(); // ← never returns
                let _ = a.grad(&inner_grads);
                eprintln!("inner backward returned (bug is fixed)");
                grad
            });
        }
    }

    let TensorPrimitive::Float(xp) = x.into_primitive() else {
        unreachable!("float input")
    };
    let out = xp.primitive.clone();
    let out = match NestedIdentity
        .prepare::<burn::backend::autodiff::checkpoint::strategy::NoCheckpointing>([xp
            .node
            .clone()])
        .compute_bound()
        .stateful()
    {
        OpsKind::Tracked(prep) => prep.finish((), out),
        OpsKind::UnTracked(prep) => prep.finish(out),
    };
    Tensor::from_primitive(TensorPrimitive::Float(out))
}

fn main() {
    // The deadlock is permanent; a watchdog converts the hang into a
    // diagnosable non-zero exit.
    std::thread::spawn(|| {
        std::thread::sleep(std::time::Duration::from_secs(15));
        eprintln!("DEADLOCK: outer backward did not complete within 15 s");
        std::process::exit(1);
    });

    let x = Tensor::<Ad, 1>::ones([4], &Default::default()).require_grad();
    let y = nested_identity(x.clone());
    let loss = y.sum();
    eprintln!("starting outer backward…");
    let grads = loss.backward();
    let gx = x.grad(&grads).expect("grad for x");
    println!("outer backward completed; grad = {gx}");
}
```

Observed output:

```
starting outer backward…
entering nested inner backward…
DEADLOCK: outer backward did not complete within 15 s
```

Expected: the inner backward completes (its graph is disjoint) and the program prints the outer gradient.

## Suggested fix (verified)

Make `cleanup_orphaned_entries` use `try_lock()` and skip graphs whose mutex is contended. The sweep is best-effort garbage collection — a graph that is busy right now is by definition still in use, and any genuinely orphaned entries it holds will be collected by the sweep at the end of a later `backward()` call:

```diff
         for graph in graphs.values() {
             {
-                let mut guard = graph.state.lock();
+                // Best-effort sweep: a graph whose mutex is contended is in
+                // use (possibly by this very thread, mid-backward) — skip it;
+                // a later backward's sweep will collect its orphans.
+                let Some(mut guard) = graph.state.try_lock() else {
+                    continue;
+                };
                 // Double safety: in case it was marked as no longer useful, but other
```

Verified: with this patch applied to a v0.21.0 checkout, the reproduction below completes instead of hanging — the inner backward returns and the outer backward produces the correct gradient (`[1.0, 1.0, 1.0, 1.0]`). A PR with this change follows.

Alternatives (not preferred here): releasing the graph state lock around each step's execution, or documenting nested backward as unsupported and failing loudly instead of hanging; #5043's `Backward`-trait rework (`&self` steps for `backward_retain`) may also be a natural place to make nested backward first-class longer-term.

Related: #4075 / #4342 / #4345 reworked this same orphan-cleanup path; #3123 fixed a prior lock-capture deadlock in the same family.

---
Observed while implementing block-level gradient checkpointing for a LoRA trainer (worked around by restructuring into sequential standalone graphs so no backward ever runs inside another). Repro executed against crates.io `burn 0.21.0`; locking path inspected at `main@e5467f0`; fix verified against a patched `v0.21.0` checkout — all on 2026-07-19.
